pkg/apis/core/validation/validation_declarative.go

```diff
@@ -0,0 +1,94 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/klog/v2"
+
+	fldtest "k8s.io/apimachinery/pkg/util/validation/field/testing"
+	validationmetrics "k8s.io/component-base/metrics/prometheus/validation"
+)
+
+// gatherDeclarativeValidationMismatches compares imperative and declarative validation errors
+// and returns detailed information about any mismatches found.  Errors are compared via type, field, and origin
+func gatherDeclarativeValidationMismatches(imperativeErrs, declarativeErrs field.ErrorList) []string {
+	// Only compare imperative errors that are actually CoveredByDeclarative.
+	coveredImperative := make(field.ErrorList, 0, len(imperativeErrs))
+	for _, err := range imperativeErrs {
+		if err.CoveredByDeclarative {
+			coveredImperative = append(coveredImperative, err)
+		}
+	}
+
+	matcher := fldtest.ErrorMatcher{}.ByType().ByField().ByOrigin().RequireOriginWhenInvalid()
+
+	matchedDeclarative := make([]bool, len(declarativeErrs))
+	var mismatchDetails []string
+
+	// Match each "covered" imperative error to a declarative error.
+	for _, iErr := range coveredImperative {
```

Comment:
Why accumulate coveredImperative and then loop a second time instead of gating with `if !err.CoveredByDeclarative { continue}` in a single loop?

Reply:
Updated, this logic is now in a single loop now